src/goto-programs/validate_goto_model.cpp

```diff
@@ -0,0 +1,251 @@
+
+Author: Diffblue
+
+Date: Oct 2018
+
+\*******************************************************************/
+#include "validate_goto_model.h"
+
+#include <set>
+
+#include "goto_functions.h"
+#include <util/invariant.h>
+
+class validate_goto_modelt
+{
+public:
+  using function_mapt = goto_functionst::function_mapt;
+
+  validate_goto_modelt(
+    const goto_functionst &goto_functions,
+    const validation_modet vm)
+    : vm{vm}, function_map{goto_functions.function_map}
+  {
+  }
+
+  void
+  do_goto_program_checks(goto_model_validation_optionst validation_options);
+
+private:
+  /// Check the goto_program has an entry point
+  ///
+  /// NB goto-cc -c will generate goto-programs without entry-points
+  /// until they are linked.
+  void entry_point_exists();
+
+  /// Check that no functions calls via function pointer are present
+  void function_pointer_calls_removed();
+
+  /// Check returns have been removed
+  ///
+  /// NB Calls via function pointer must have been removed already when
+  /// removing returns, thus enabling this check also should enable the check
+  /// that all calls via function pointer have been removed
+  ///
+  /// Sub-checks are:
+  /// - no return statements in any of the functions
+  /// - lhs of every \ref code_function_callt instruction is nil
+  /// - all return types are void (of both calls and functions themselves)
+  void check_returns_removed();
+
+  /// Check that for all
+  /// 1. functions that are called or
+  /// 2. functions of which the address is taken,
+  /// a corresponding entry exists in the function_map
+  void check_called_functions();
+
+  /// Check that goto-programs that have a body end with an END_FUNCTION
+  /// instruction.
+  ///
+  /// NB functions that are only declared, and not used will not appear in the
+  /// function map. Functions that are declared only and used to e.g.
+  /// initialise a function pointer will have no body.
+  void check_last_instruction();
+
+  /// Check both
+  /// 1. every instruction \"code\" field, has a non nil source location"
+  /// 2. every instruction source_location field is nil
+  /// NB this check is not expected to pass until setting these locations is
+  /// patched up.
+  void check_sourcecode_location();
+
+  const validation_modet vm;
+  const function_mapt &function_map;
+  const goto_model_validation_optionst validation_options;
+};
+
+void validate_goto_modelt::do_goto_program_checks(
+  goto_model_validation_optionst validation_options)
+{
+  if(validation_options.entry_point_exists)
+    entry_point_exists();
+
+  if(validation_options.check_last_instruction)
+    check_last_instruction();
+
+  /// NB function pointer calls must have been removed before removing
+  /// returns - so 'check_returns_removed' also enables
+  // 'function_pointer_calls_removed'
+  if(
+    validation_options.function_pointer_calls_removed ||
+    validation_options.check_returns_removed)
+    function_pointer_calls_removed();
```

Comment:
Please use braces when the condition spans multiple lines.

Reply:
Done.